File: tools/fnt.py
#!/usr/bin/env python3
"""NE `.FON` (Windows 2.x/3.x FNT bitmap font) 解析器。

WinCV 隨附四個 `.FON`,face name 分別是 cvga / cvga1018 / cvga1224
(WinCV.fon 與 cvga.fon 同名同尺寸)。程式以 AddFontResource 註冊它們,
再用 CreateFontIndirect 指名 face 取得半形字型。

FNT 2.0 header 版面(bytes,little-endian):
    0x00 dfVersion  W    0x42 dfType      W    0x55 dfCharSet   B
    0x02 dfSize     D    0x44 dfPoints    W    0x56 dfPixWidth  W
    0x06 dfCopyright 60  0x46 dfVertRes   W    0x58 dfPixHeight W
                         0x48 dfHorizRes  W    0x5A dfPitchAndFamily B
                         0x4A dfAscent    W    0x5B dfAvgWidth  W
                         0x4C dfIntLeading W   0x5D dfMaxWidth  W
                         0x4E dfExtLeading W   0x5F dfFirstChar B
                         0x50 dfItalic    B    0x60 dfLastChar  B
                         0x51 dfUnderline B    0x61 dfDefaultChar B
                         0x52 dfStrikeOut B    0x62 dfBreakChar B
                         0x53 dfWeight    W    0x63 dfWidthBytes W
                                               0x65 dfDevice    D
                                               0x69 dfFace      D
                                               0x71 dfBitsOffset D
    0x76 dfCharTable: (dfLastChar - dfFirstChar + 2) 筆 {W width, W offset}

字模是 **column-major**:每個字元先存第 0 欄的全部列(每 8 列一個 byte),
再第 1 欄……寬度 > 8 的字元因此分成多個 8-pixel 欄組。

用法:
  python3 tools/fnt.py info  original/app/cvga.fon
  python3 tools/fnt.py atlas original/app/cvga.fon out.pbm
  python3 tools/fnt.py glyph original/app/cvga.fon 0x41
"""
import struct
import sys


def ne_resources(d):
    ne = struct.unpack_from("<H", d, 0x3C)[0]
    rt = ne + struct.unpack_from("<H", d, ne + 0x24)[0]
    shift = struct.unpack_from("<H", d, rt)[0]
    p = rt + 2
    out = []
    while struct.unpack_from("<H", d, p)[0] != 0:
        tid = struct.unpack_from("<H", d, p)[0]
        cnt = struct.unpack_from("<H", d, p + 2)[0]
        p += 8
        for _ in range(cnt):
            off, ln = struct.unpack_from("<HH", d, p)
            out.append((tid & 0x7FFF, off << shift, ln << shift))
            p += 12
    return out


class Fnt:
    def __init__(self, d, base):
        self.d = d
        self.base = base
        u16 = lambda o: struct.unpack_from("<H", d, base + o)[0]
        u32 = lambda o: struct.unpack_from("<I", d, base + o)[0]
        self.version = u16(0x00)
        self.size = u32(0x02)
        self.points = u16(0x44)
        self.ascent = u16(0x4A)
        self.int_leading = u16(0x4C)
        self.ext_leading = u16(0x4E)
        self.weight = u16(0x53)
        self.charset = d[base + 0x55]
        self.pix_width = u16(0x56)
        self.pix_height = u16(0x58)
        self.pitch_family = d[base + 0x5A]
        self.avg_width = u16(0x5B)
        self.max_width = u16(0x5D)
        self.first = d[base + 0x5F]
        self.last = d[base + 0x60]
        self.default_char = d[base + 0x61]
        self.break_char = d[base + 0x62]
        self.width_bytes = u16(0x63)
        face_off = u32(0x69)
        end = d.find(b"\0", base + face_off)
        self.face = d[base + face_off:end].decode("latin1")

        n = self.last - self.first + 2
        self.chars = []
        for i in range(n):
            w, off = struct.unpack_from("<HH", d, base + 0x76 + i * 4)
            self.chars.append((w, off))

    def glyph(self, code):
        """回傳 list[str],每列一個 '.#' 字串。"""
        i = code - self.first
        if not (0 <= i < len(self.chars) - 1):
            return None
        w, off = self.chars[i]
        h = self.pix_height
        cols = (w + 7) // 8
        rows = ["" for _ in range(h)]
        for c in range(cols):
            colbase = self.base + off + c * h
            for y in range(h):
                b = self.d[colbase + y]
                for bit in range(8):
                    if c * 8 + bit < w:
                        rows[y] += "#" if b & (0x80 >> bit) else "."
        return rows


def load(path):
    d = open(path, "rb").read()
    for tid, off, _ln in ne_resources(d):
        if tid == 8:  # RT_FONT
            return Fnt(d, off)
    raise SystemExit("找不到 RT_FONT resource")


def cmd_info(f):
    print(f'face          "{f.face}"')
    print(f"version       {f.version:#06x}")
    print(f"points        {f.points}")
    print(f"pixel W x H   {f.pix_width} x {f.pix_height}   (0 = 變寬)")
    print(f"avg / max W   {f.avg_width} / {f.max_width}")
    print(f"ascent        {f.ascent}  int_leading {f.int_leading}  ext {f.ext_leading}")
    print(f"charset       {f.charset:#04x}  pitch_family {f.pitch_family:#04x}")
    print(f"chars         {f.first:#04x} - {f.last:#04x}  default {f.default_char:#04x}")
    widths = {}
    for i in range(f.last - f.first + 1):
        widths[f.chars[i][0]] = widths.get(f.chars[i][0], 0) + 1
    print(f"寬度分布       {dict(sorted(widths.items()))}")


def cmd_glyph(f, code):
    g = f.glyph(code)
    if g is None:
        raise SystemExit(f"{code:#04x} 不在字型範圍")
    print(f"{code:#04x} width={f.chars[code - f.first][0]}")
    for r in g:
        print(r)


def cmd_atlas(f, out):
    """16 x 16 排的字模圖,存成 PBM(不需要任何影像函式庫)。"""
    cw = max(w for w, _ in f.chars[:f.last - f.first + 1]) or 8
    ch = f.pix_height
    W, H = cw * 16, ch * 16
    bits = [[0] * W for _ in range(H)]
    for code in range(f.first, f.last + 1):
        g = f.glyph(code)
        if not g:
            continue
        gx, gy = (code % 16) * cw, (code // 16) * ch
        for y, row in enumerate(g):
            for x, c in enumerate(row):
                if c == "#":
                    bits[gy + y][gx + x] = 1
    with open(out, "wb") as fh:
        fh.write(b"P1\n%d %d\n" % (W, H))
        for row in bits:
            fh.write(("".join(str(b) for b in row) + "\n").encode())
    print(f"{out}  {W}x{H}  cell {cw}x{ch}")


def main():
    if len(sys.argv) < 3:
        sys.exit(__doc__)
    cmd, path = sys.argv[1], sys.argv[2]
    f = load(path)
    if cmd == "info":
        cmd_info(f)
    elif cmd == "glyph":
        cmd_glyph(f, int(sys.argv[3], 0))
    elif cmd == "atlas":
        cmd_atlas(f, sys.argv[3])
    else:
        sys.exit(__doc__)


if __name__ == "__main__":
    main()

File: internal/fnt/fnt.go
// Package fnt 讀 Windows 2.x/3.x 的 NE `.FON` 點陣字型。
//
// WinCV 隨附 cvga (8x15) / cvga1018 (10x18) / cvga1224 (12x24) 三個半形字型,
// 三者都是定寬、涵蓋 0x00-0xFF。原版以 AddFontResource 註冊後用 face 名指定。
//
// 格式細節與各欄位的 offset 見 tools/fnt.py 的檔頭註解;兩邊是同一份規格,
// Python 版拿來探索與驗證,這裡是執行期用的。
package fnt

import (
	"encoding/binary"
	"fmt"
	"github.com/wicanr2/wincv-remake/internal/i18n"
)

// Font 是一個 .FON 裡的單一字型。
type Font struct {
	Face      string
	PixWidth  int // 0 表示變寬
	PixHeight int
	Ascent    int
	First     byte
	Last      byte
	widths    []int
	offsets   []int
	data      []byte
	base      int
}

// Size 回傳格子的字身大小。實作 render.HalfSource。
func (f *Font) Size() (w, h int) { return f.PixWidth, f.PixHeight }

// Glyph 是一個字的點陣圖,以 row-major 的 bool 表示。
type Glyph struct {
	W, H int
	Bits []bool // len == W*H,true 表示前景
	// Alpha 是每個像素的覆蓋率(0–255),有的話畫的時候用它混色,
	// 沒有(nil)就是純 1-bit 的點陣字。向量字型在大字級時用這個
	// 做反鋸齒;點陣字永遠是 nil —— 它的邊緣本來就該是硬的。
	Alpha []uint8
}

// Cov 回傳 (x, y) 的覆蓋率。1-bit 字模回 0 或 255。
func (g *Glyph) Cov(x, y int) uint8 {
	if x < 0 || y < 0 || x >= g.W || y >= g.H {
		return 0
	}
	if g.Alpha != nil {
		return g.Alpha[y*g.W+x]
	}
	if g.Bits[y*g.W+x] {
		return 255
	}
	return 0
}

func (g *Glyph) At(x, y int) bool {
	if x < 0 || y < 0 || x >= g.W || y >= g.H {
		return false
	}
	return g.Bits[y*g.W+x]
}

func u16(d []byte, o int) int { return int(binary.LittleEndian.Uint16(d[o:])) }
func u32(d []byte, o int) int { return int(binary.LittleEndian.Uint32(d[o:])) }

// neFontResources 找出 NE 檔裡所有 RT_FONT (type 8) 的位置。
func neFontResources(d []byte) ([]int, error) {
	if len(d) < 0x40 || d[0] != 'M' || d[1] != 'Z' {
		return nil, fmt.Errorf(i18n.T("不是 MZ 檔"))
	}
	ne := u16(d, 0x3C)
	if ne+0x26 > len(d) || d[ne] != 'N' || d[ne+1] != 'E' {
		return nil, fmt.Errorf(i18n.T("找不到 NE header"))
	}
	rt := ne + u16(d, ne+0x24)
	if rt+2 > len(d) {
		return nil, fmt.Errorf(i18n.T("resource table 超出檔案"))
	}
	shift := u16(d, rt)
	p := rt + 2
	var out []int
	for p+8 <= len(d) {
		tid := u16(d, p)
		if tid == 0 {
			break
		}
		cnt := u16(d, p+2)
		p += 8
		for i := 0; i < cnt && p+12 <= len(d); i++ {
			off := u16(d, p) << shift
			if tid&0x7FFF == 8 {
				out = append(out, off)
			}
			p += 12
		}
	}
	if len(out) == 0 {
		return nil, fmt.Errorf(i18n.T("沒有 RT_FONT resource"))
	}
	return out, nil
}

// Parse 讀一個 .FON 的位元組,回傳裡面第一個字型。
func Parse(d []byte) (*Font, error) {
	offs, err := neFontResources(d)
	if err != nil {
		return nil, err
	}
	return parseAt(d, offs[0])
}

func parseAt(d []byte, base int) (*Font, error) {
	if base+0x76 > len(d) {
		return nil, fmt.Errorf(i18n.T("FNT header 超出檔案"))
	}
	f := &Font{
		data:      d,
		base:      base,
		Ascent:    u16(d, base+0x4A),
		PixWidth:  u16(d, base+0x56),
		PixHeight: u16(d, base+0x58),
		First:     d[base+0x5F],
		Last:      d[base+0x60],
	}
	faceOff := base + u32(d, base+0x69)
	if faceOff >= len(d) {
		return nil, fmt.Errorf(i18n.T("dfFace 超出檔案"))
	}
	end := faceOff
	for end < len(d) && d[end] != 0 {
		end++
	}
	f.Face = string(d[faceOff:end])

	n := int(f.Last) - int(f.First) + 2
	if base+0x76+n*4 > len(d) {
		return nil, fmt.Errorf(i18n.T("dfCharTable 超出檔案"))
	}
	f.widths = make([]int, n)
	f.offsets = make([]int, n)
	for i := 0; i < n; i++ {
		f.widths[i] = u16(d, base+0x76+i*4)
		f.offsets[i] = u16(d, base+0x76+i*4+2)
	}
	return f, nil
}

// Glyph 取一個字的點陣圖。字模是 column-major:先存第 0 欄的全部列
// (每 8 列一個 byte),再第 1 欄;寬度超過 8 的字因此分成多個 8-pixel 欄組。
// 不在範圍內回 nil。
func (f *Font) Glyph(code byte) *Glyph {
	if code < f.First || code > f.Last {
		return nil
	}
	i := int(code) - int(f.First)
	w, off := f.widths[i], f.offsets[i]
	h := f.PixHeight
	if w == 0 || h == 0 {
		return nil
	}
	g := &Glyph{W: w, H: h, Bits: make([]bool, w*h)}
	cols := (w + 7) / 8
	for c := 0; c < cols; c++ {
		colBase := f.base + off + c*h
		for y := 0; y < h; y++ {
			if colBase+y >= len(f.data) {
				return g
			}
			b := f.data[colBase+y]
			for bit := 0; bit < 8; bit++ {
				x := c*8 + bit
				if x >= w {
					break
				}
				if b&(0x80>>uint(bit)) != 0 {
					g.Bits[y*w+x] = true
				}
			}
		}
	}
	return g
}
